### DISCUSSION TITLE
Is there a way to get access_token (jwt) for a user who will authenticate using google OAuth2?

### DISCUSSION THREAD

**Discussion by vivekko** [keycloak/keycloak#22614] [Q&A]
I have figured out way using rest api to get access_token for a user who authenticates via username password but I want access_token for those who authenticated via google OAuth2. I have searched a lot and couldn't find anything useful. Can you recommend me ways to do it ?

**Comment by justin-tay** [comment]
I'm assuming that you want Keycloak to broker to Google. 

If that is the case you would need to configure `Identity providers > Add provider > Google`

The tutorial at https://keycloakthemes.com/blog/how-to-setup-sign-in-with-google-using-keycloak seems to cover the essential steps.

Note that what you get at the end here is tokens issued by Keycloak and not tokens issued by Google since it's not clear from your question whether that is what you want.

**Comment by vivekko** [comment]
Hi @justin-tay 
I have already completed the configuration for google as a identity provider and it works as well. But I am not sure how to get the access_token for the user who used google authentication. 

I am talking about the access_token which we get using this endpoint issued by keycloak: 
POST /realms/master/protocol/openid-connect/token

**Reply by justin-tay** [reply]
There's no difference getting the tokens from the Keycloak endpoint whether or not it brokered Google or if they used their username and password to login to Keycloak.

If you are using Standard flow (Authorization Code Flow) your backend (Relying Party) will get the `code` from the browser redirect from Keycloak to pass to the token endpoint.

You can look at this part of the [tutorial](https://www.appsdeveloperblog.com/keycloak-authorization-code-grant-example/#Exchanging_Authorization_Code_for_Access_Token) on exchanging the authorization code for the access token.

**Reply by Selva-Torus** [reply]
I found the REST API for google/github oauth provider authentication access_token from https://www.appsdeveloperblog.com/keycloak-authorization-code-grant-example/#Exchanging_Authorization_Code_for_Access_Token
this link provided by you , but however I've had a complexity in getting that Oauth authorization code from that provider ,

I use Next.js project and in this I got that code from the url , but as soon as the url called in the server's terminal callback happened without getting token, Note : I made all the setup of identity provider in keycloak , Now I call this oauth authorization via RestAPI , 

please suggest me how to make it work

**Reply by ishaan812** [reply]
did anyone figure out how to make htis work without token exchange

**Comment by ZIRAKrezovic** [comment]
Since this seemed to pop up lately, there is a section in the server documentation https://www.keycloak.org/docs/latest/server_admin/index.html#retrieving-external-idp-tokens
